Mathlib/Combinatorics/SimpleGraph/Metric.lean

```diff
@@ -43,105 +41,179 @@ variable {V : Type*} (G : SimpleGraph V)
 
 /-! ## Metric -/
 
+section edist
+
+/--
+The extended distance between two vertices is the length of the shortest walk between them.
+It is `⊤` if no such walk exists.
+-/
+noncomputable def edist (u v : V) : ℕ∞ :=
+  ⨅ w : G.Walk u v, w.length
+
+variable {G} {u v w : V}
+
+theorem edist_eq_sInf : G.edist u v = sInf (Set.range fun w : G.Walk u v ↦ (w.length : ℕ∞)) := rfl
+
+protected theorem Reachable.exists_walk_length_eq_edist (hr : G.Reachable u v) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  csInf_mem <| Set.range_nonempty_iff_nonempty.mpr hr
 
-/-- The distance between two vertices is the length of the shortest walk between them.
-If no such walk exists, this uses the junk value of `0`. -/
+protected theorem Connected.exists_walk_length_eq_edist  (hconn : G.Connected) (u v : V) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (hconn u v).exists_walk_length_eq_edist
+
+theorem edist_le (p : G.Walk u v) :
+    G.edist u v ≤ p.length :=
+  sInf_le ⟨p, rfl⟩
+protected alias Walk.edist_le := edist_le
+
+@[simp]
+theorem edist_eq_zero_iff :
+    G.edist u v = 0 ↔ u = v := by
+  apply Iff.intro <;> simp [edist, ENat.iInf_eq_zero]
+
+theorem edist_self : edist G v v = 0 :=
+  edist_eq_zero_iff.mpr rfl
+
+theorem edist_pos_of_ne (hne : u ≠ v) :
+    0 < G.edist u v :=
+  pos_iff_ne_zero.mpr <| edist_eq_zero_iff.ne.mpr hne
+protected alias Ne.edist_pos := edist_pos_of_ne
+
+lemma edist_eq_top_of_not_reachable (h : ¬G.Reachable u v) :
+    G.edist u v = ⊤ := by
+  simp [edist, not_reachable_iff_isEmpty_walk.mp h]
+
+theorem reachable_of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    G.Reachable u v :=
+  not_not.mp <| edist_eq_top_of_not_reachable.mt h
+
+protected theorem Connected.edist_triangle (hconn : G.Connected) :
```

Comment:
This doesn't need connectedness since we're using edist!
```
protected theorem edist_triangle : G.edist u w ≤ G.edist u v + G.edist v w := by
  rcases eq_or_ne (G.edist u v) ⊤ with huv | huv
  case inl => simp [huv]
  case inr =>
    rcases eq_or_ne (G.edist v w) ⊤ with hvw | hvw
    case inl => simp [hvw]
    case inr =>
      obtain ⟨p, hp⟩ := exists_walk_of_edist_ne_top huv
      obtain ⟨q, hq⟩ := exists_walk_of_edist_ne_top hvw
      rw [←hp, ←hq, ←Nat.cast_add, ←Walk.length_append]
      exact edist_le _
 ```

Reply:
It's probably not worth developing all this theory, but it occurred to me that we could pull back infinity to Walk to have "extended walks", along with the "does not exist" walk. Then `edist_triangle` could be proved without any cases, because it's already been done in `EWalk.length_append`.
```lean
def EWalk (u v : V) := Option (G.Walk u v)

def EWalk.length : G.EWalk u v → ℕ∞ := Option.map Walk.length

def EWalk.append (p : G.EWalk u v) (q : G.EWalk v w) : G.EWalk u w :=
  Walk.append <$> p <*> q

def EWalk.length_append {p : G.EWalk u v} {q : G.EWalk v w} :
    (p.append q).length = p.length + q.length := by
  cases p <;> cases q <;> simp [EWalk.append, EWalk.length] <;> rfl

theorem exists_ewalk_length_eq_edist :
    ∃ p : G.EWalk u v, p.length = G.edist u v := by
  sorry
```

Reply:
It is a great idea! I'll keep it in mind to potentially work on.
I think it would be nicer to move `Walk`s from `Connectivity.lean` into their own file (in the `Connectivity` folder) before attempting that, what do you think?

Reply:
I think that change does sound valuable - Reachable and Connected can be in Connectivity.lean; Walks and Paths can be in Walk.lean

Reply:
#15171 does this.